### PR TITLE
Add support for shutting down vms

### PR DESCRIPTION
**Pull request by vigejolla**

### CODE REVIEW

**Review by martyone — COMMENTED** [review]
Check for style issues :)

**Comment by vigejolla** [comment]
> Check for style issues :)

Could you be a bit more specific, as I don't see any style issues...

**Comment by martyone** [comment]
Try harder :)

**Comment by vigejolla** [comment]
Could you be a bit more disparaging?

**Comment by martyone** [comment]
It makes me feel degraded to a code formatting tool when I need to provide this kind of feedback under every PR :) https://wiki.qt.io/Qt_Coding_Style#Braces

**Comment by vigejolla** [comment]
I found one { on a wrong place. Being arrogant does not help me learn these kind of things any faster. If you know what's wrong with my code just tell me what it is and not keep me guessing. I guess I'll start writing comments to your PRs saying things like "I found a bug in your code, it's there somewhere, go find it!".

**Comment by martyone** [comment]
I am sure I didn't use that tone, I am sure we all know which documents we follow and I believe I mentioned these already. So while you perceived me as being arrogant, which I am sorry about, I was just trying to do the best to "teach a man to fish". The wrong positioned brace is not the only issue with braces :(

**Comment by vigejolla** [comment]
You may think you are doing your best to teach, but I'm sure you can do better. I can't fix issues if you don't tell me what they are.

**Comment by vigejolla** [comment]
My next guess: Those one-line if blocks which had braces. My (not so good) defense: They were copy pasted from elevatedexecuteoperation.cpp...

**Comment by martyone** [comment]
Yup. Unless you think it adds some value to keeps these two sources in sync (I doubt) I would prefer cleaning those badly positioned and extra braces.

**Comment by vigejolla** [comment]
> Yup. Unless you think it adds some value to keeps these two sources in sync (I doubt) I would prefer cleaning those badly positioned and extra braces.

As you can see I already cleaned them :)

**Comment by martyone** [comment]
Hm, for some reason I cannot see the change. Did you push it properly?

**Comment by vigejolla** [comment]
I thought I did, but it looks like I've messed up at some point...

**Comment by vigejolla** [comment]
'Check now :)

File: src/libs/installer/vmshutdownoperation.cpp [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+        else
+            args.append(argument);
+    }
+
+    return runNow ? d->run(args) : true;
+}
+
+bool VmShutdownOperation::undoOperation()
+{
+    QStringList args;
```
Comment (martyone) [inline src/libs/installer/vmshutdownoperation.cpp]: IIUC, I would maybe name it `virtualMachines`.

File: src/libs/installer/vmshutdownoperation.cpp [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+{
+    emit cancelProcess();
+}
+
+QString VmShutdownOperation::Private::vBoxManagePath()
+{
+    static QString path;
+    if (!path.isNull()) {
+        return path;
+    }
```
Comment (martyone) [inline src/libs/installer/vmshutdownoperation.cpp]: here

File: src/libs/installer/vmshutdownoperation.cpp [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+{
+    std::unique_ptr<QProcessWrapper> process(new QProcessWrapper());
+
+    connect(q, &VmShutdownOperation::cancelProcess, process.get(), &QProcessWrapper::cancel);
+
+    //we still like the none blocking possibility to perform this operation without threads
+    QEventLoop loop;
+    if (QThread::currentThread() == qApp->thread()) {
+        QObject::connect(process.get(), &QProcessWrapper::finished, &loop, &QEventLoop::quit);
+    }
```
Comment (martyone) [inline src/libs/installer/vmshutdownoperation.cpp]: here

File: src/libs/installer/vmshutdownoperation.cpp [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+
+    process->start(vBoxManagePath(), args);
+
+    bool success = false;
+    //we still like the none blocking possibility to perform this operation without threads
+    if (QThread::currentThread() == qApp->thread()) {
+        success = process->waitForStarted();
+    } else {
+        success = process->waitForFinished(-1);
+    }
```
Comment (martyone) [inline src/libs/installer/vmshutdownoperation.cpp]: here

File: src/libs/installer/vmshutdownoperation.cpp [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+    if (!success) {
+        q->setError(UserDefinedError);
+        q->setErrorString(tr("Cannot start VBoxManage: %2").arg(process->errorString()));
+        return false;
+    }
+
+    if (QThread::currentThread() == qApp->thread()) {
+        if (process->state() != QProcessWrapper::NotRunning) {
+            loop.exec();
+        }
```
Comment (martyone) [inline src/libs/installer/vmshutdownoperation.cpp]: here

File: src/libs/installer/vmshutdownoperation.cpp [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+}
+
+bool VmShutdownOperation::Private::isVmRunning(const QString &vboxName)
+{
+    QStringList args;
+    args << QLatin1String("showvminfo") << vboxName;
+    QString output;
+    if (!executeVboxCommand(args, &output)) {
+        return true;
+    }
```
Comment (martyone) [inline src/libs/installer/vmshutdownoperation.cpp]: here

File: src/libs/installer/vmshutdownoperation.cpp [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+        return false;
+
+    QTime timeout;
+    timeout.start();
+
+    while (timeout.elapsed() < 30000 ) {
+        if (!isVmRunning(args.first()))
+        {
+            return true;
+        }
```
Comment (martyone) [inline src/libs/installer/vmshutdownoperation.cpp]: here

File: src/libs/installer/vmshutdownoperation.cpp [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+public:
+    explicit Private(VmShutdownOperation* qq)
+        : q(qq)
+    {
+    }
+
+private:
+    QString vBoxManagePath();
+    bool executeVboxCommand(const QStringList& args, QString* output = nullptr);
+    bool isVmRunning(const QString& vboxName);
```
Comment (martyone) [inline src/libs/installer/vmshutdownoperation.cpp]: rather `vmName`?

File: src/libs/installer/vmshutdownoperation.cpp [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+    }
+    QRegularExpression re(QStringLiteral("^Session (name|type):"), QRegularExpression::MultilineOption);
+    return re.match(output).hasMatch();
+}
+
+bool VmShutdownOperation::Private::run(const QStringList& args)
+{
+    // it is ok to have no arguments, if vm is only shutdown
+    // during undo
+    if (args.isEmpty())
```
Comment (martyone) [inline src/libs/installer/vmshutdownoperation.cpp]: What if more args are passed? Accepting just one VM name would be OK with me.

**Review by martyone — COMMENTED** [review]
There must be something wrong, as I still do not see them updated - see the "here" comments.

**Comment by vigejolla** [comment]
> There must be something wrong, as I still do not see them updated - see the "here" comments.

It seems I'm unbelievable bad at using git, sorry.

File: src/libs/installer/vmshutdownoperation.cpp [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+        }
+    }
+#endif // Q_OS_WIN
+
+    return path;
+}
+
+bool VmShutdownOperation::Private::executeVboxCommand(const QStringList &args, QString *output)
+{
+    std::unique_ptr<QProcessWrapper> process(new QProcessWrapper());
```
Comment (martyone) [inline src/libs/installer/vmshutdownoperation.cpp:174]: You should use `auto process = std::make_unique...`. The point is to be able to find all raw new/delete usage just by grepping the source.

Comment (vigejolla) [reply src/libs/installer/vmshutdownoperation.cpp:174]: Unfortunately the installer framework is compiled with -std=gnu++11. There is no make_unique available.

Comment (martyone) [reply src/libs/installer/vmshutdownoperation.cpp:174]: phew.. then QScopedPointer might look more native here?

**Review by martyone — APPROVED** [review]

**Pull request merged**